metrics/metrics.go

```diff
@@ -389,27 +389,71 @@ func GroupHandler() http.Handler {
 	return promhttp.HandlerFor(GroupMetrics, promhttp.HandlerOpts{Registry: GroupMetrics})
 }
 
-// lazyPeerHandler is a structure that defers learning who current
-// group members are until an http request is received.
+// lazyPeerHandler is a structure that defers learning who current group members are until
+// an http request is received for a specific peer
 type lazyPeerHandler struct {
-	peerHandler PeerHandler
+	groupHandlers []GroupHandlers
+	// peerHandlers is a cache of peer address -> handler
+	// TODO Do we need to evict from cache at some point? The only case when it should be
+	//      invalidated is if a peer leaves e.g. during a Reshare
+	peerHandlers sync.Map
+}
+
+// newLazyPeerHandler creates a new lazyPeerHandler from a slice of GroupHandlers
+func newLazyPeerHandler(groupHandlers []GroupHandlers) *lazyPeerHandler {
+	return &lazyPeerHandler{
+		groupHandlers,
+		sync.Map{},
+	}
 }
 
+// handlerForPeer returns the http.Handler associated with a given peer address.
+// If a peer belongs to multiple groups, it will return the handler associated with the
+// one that appears first in groupHandlers.
```

Comment:
this seems weird - would we not prefer to extend the metrics protocol to allow peers to request a specific group that they want metrics back for?

Reply:
I see what you mean in that it feels a bit weird, however,the metrics are group-agnostic (they contain metrics for all groups that the node is a part of). If we made it so that you could request for a specific group, we would then have to modify deployment in `drand-infra` each time we add a new beacon. What we want here is to be able to get the metrics for all nodes that this node is connected to, regardless of the group.

We could refactor it so that we have a global, non-group-dependent GRPC connection for the sole purpose of sending metrics (and possibly other control messages in the future), but I'm not sure that that would be worth it at this point.

Reply:
Another option I thought of over the weekend, is to populate a `sync.Map` whenever we try to connect to a host. I believe that would be here: https://github.com/drand/drand/blob/master/net/client_grpc.go#L325. We could add something like this after the `if err != nil { ... }` block, assuming that `metrics.Handlers` is a `sync.Map`:

```go
metrics.Handlers.Store(p.Address(), g.HandleHTTP(p)
```

we would then entirely remove`MetricsHandlerForPeer` and `handlerForPeer`, and replace the call to `handlerForPeer` here https://github.com/drand/drand/pull/1017/files#diff-d715f391f2ffa038c3167bfea1adf741c1f5e39f08349fbbacab780a594146a4R453-R460 with:

```go
handler, found := Handlers.Load(addr)
if !found {
  log.DefaultLogger().Warnw("", "metrics", "peer not found", "addr", addr)
  w.WriteHeader(http.StatusNotFound)
  return
}
```

It does feel much cleaner that way, WDYT?